solr/src/main/scala/akka/stream/alpakka/solr/SolrFlowStage.scala

```diff
@@ -4,179 +4,375 @@
+          doc.addField(field, jMap)
+        }
       }
-    } else {
-      sendBulkToSolr(nextMessages)
+      doc
     }
+    if (log.isDebugEnabled) log.debug(s"Update atomically $docs")
+    client.add(collection, docs.asJava, settings.commitWithin)
   }
 
-  private def handleSuccess(): Unit =
-    completeStage()
+  private def deleteBulkToSolrByIds(messages: Seq[IncomingMessage[T, C]]): UpdateResponse = {
+    val docsIds = messages
+      .filter { message =>
+        message.operation == DeleteByIds && message.idFieldValueOpt.isDefined
+      }
+      .map { message =>
+        message.idFieldValueOpt.get
+      }
+    if (log.isDebugEnabled) log.debug(s"Delete the ids $docsIds")
+    client.deleteById(collection, docsIds.asJava, settings.commitWithin)
+  }
 
-  private def sendBulkToSolr(messages: Seq[IncomingMessage[T, C]]): Unit =
-    try {
-      val docs = messages.map(message => messageBinder(message.source))
-      val response = client.add(collection, docs.asJava, settings.commitWithin)
-      handleResponse(messages, response.getStatus)
-    } catch {
-      case NonFatal(exc) =>
-        val rootCause = SolrException.getRootCause(exc)
-        if (shouldRetry(rootCause)) {
-          handleFailure(messages, exc)
-        } else {
-          val status = exc match {
-            case e: SolrException => e.code()
-            case _ => -1
-          }
-          handleResponse(messages, status)
-        }
+  private def deleteEachByQuery(messages: Seq[IncomingMessage[T, C]]): UpdateResponse = {
+    val responses = messages.map { message =>
+      val query = message.queryOpt.get
+      if (log.isDebugEnabled) log.debug(s"Delete by the query $query")
+      client.deleteByQuery(collection, query, settings.commitWithin)
     }
+    responses.filter(r => r.getStatus != 0).headOption.getOrElse(responses.head)
+  }
 
-  private def shouldRetry(cause: Throwable): Boolean =
-    cause match {
-      case _: ConnectException => true
-      case _: NoHttpResponseException => true
-      case _: SocketException => true
-      case _ => false
+  private def sendBulkToSolr(messages: Seq[IncomingMessage[T, C]]): Unit = {
+
+    @tailrec
+    def send(toSend: Seq[IncomingMessage[T, C]]): UpdateResponse = {
+      val operation = toSend.head.operation
+      //Just take a subset of this operation
+      val current = toSend.takeWhile { m =>
+        m.operation == operation
+      }
+      //send this subset
+      val response = operation match {
+        case Upsert => updateBulkToSolr(current)
+        case AtomicUpdate => atomicUpdateBulkToSolr(current)
+        case DeleteByIds => deleteBulkToSolrByIds(current)
+        case DeleteByQuery => deleteEachByQuery(current)
+      }
+      //Now take the remaining
+      val remaining = toSend.dropWhile(m => m.operation == operation)
+      if (remaining.nonEmpty) {
+        send(remaining) //Important: Not really recursive, because the future breaks the recursion
```

Comment:
I think this comment is outdated, is it not?

Reply:
Thx